### PR TITLE
Make Dreambooth SD Training Script `torch.compile` compatible

### LINKED ISSUE #6503: [Training] Make officially maintained training scripts compatible with `torch.compile()`
We recently made the DreamBooth LoRA SDXL training script fully compatible with `torch.compile()` in the following PRs: https://github.com/huggingface/diffusers/pull/6483, https://github.com/huggingface/diffusers/pull/6511. 

Since there might be speedup benefits from `torch.compile()`, it makes sense to propagate these changes to the other training scripts too. We'd appreciate help from the community in this matter. 

Below is a list of the training scripts where we'd like to integrate this:

- DreamBooth LoRA
  - [x] [Regular](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth.py) 
  - [x] [LoRA](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora.py)
- ControlNet
  - [x] [SD](https://github.com/huggingface/diffusers/blob/main/examples/controlnet/train_controlnet.py)
  - [x] [SDXL](https://github.com/huggingface/diffusers/blob/main/examples/controlnet/train_controlnet_sdxl.py)
- Text-to-image
  - [x] [SD](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py)
  - [x] [SD LoRA](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora.py)
  - [x] [SDXL](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_sdxl.py)
  - [x] [SDXL LoRA](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora_sdxl.py)
- InstructPix2Pix
  - [x] [SD](https://github.com/huggingface/diffusers/blob/main/examples/instruct_pix2pix/train_instruct_pix2pix.py)
  - [x] [SDXL](https://github.com/huggingface/diffusers/blob/main/examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py)
- T2I Adapters
  - [x] [SDXL](https://github.com/huggingface/diffusers/blob/main/examples/t2i_adapter/train_t2i_adapter_sdxl.py)

https://github.com/huggingface/diffusers/pull/6483 and https://github.com/huggingface/diffusers/pull/6511 can serve as a reference. 

Feel free to open PRs targeting only one example at a time and tag me :)

### PR DESCRIPTION
**Pull request by rootonchair**
# What does this PR do?

Fixes #6503 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by rootonchair** [comment]
@sayakpaul this is the PR for Dreambooth Regular

File: examples/dreambooth/train_dreambooth.py [hunk 1]
```diff
@@ -107,6 +108,12 @@ def save_model_card(
         f.write(yaml + model_card)
 
 
+def unwrap_model(accelerator, model):
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth.py:111]: Let's make this inside the function as done in the other PR.

**Review by sayakpaul — APPROVED** [review]
Good job. Left just one comment.

File: examples/dreambooth/train_dreambooth.py (hunk 1 above)
Comment (rootonchair) [reply examples/dreambooth/train_dreambooth.py:111]: `log_validation` lies outside of the function. Should I put `log_validation` inside too?

Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth.py:111]: `log_validation` is fine since it's a big one. However, `unwrap_model()` isn't.

Comment (rootonchair) [reply examples/dreambooth/train_dreambooth.py:111]: `log_validation` uses `unwrap_model` too so moving `unwrap_model` inside the function will raise error. Do you have any idea how to fix this?

Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth.py:111]: We can pass the model to `log_validation` in the unwrapped state, no?

Comment (rootonchair) [reply examples/dreambooth/train_dreambooth.py:111]: I see. I have made new update

File: examples/dreambooth/train_dreambooth.py [hunk 2]
```diff
@@ -794,6 +792,7 @@ def encode_prompt(text_encoder, input_ids, attention_mask, text_encoder_use_atte
     prompt_embeds = text_encoder(
         text_input_ids,
         attention_mask=attention_mask,
+        return_dict=False,
     )
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth.py:796]:
```suggestion
    prompt_embeds = text_encoder(text_input_ids, attention_mask=attention_mask, return_dict=False)
    )
```

Will probability fit within the permissible line-length.

**Review by sayakpaul — APPROVED** [review]
Works for me. Just comment!

File: examples/dreambooth/train_dreambooth.py (hunk 2 above)
Comment (rootonchair) [reply examples/dreambooth/train_dreambooth.py:796]: Sure!

**Comment by sayakpaul** [comment]
Oops. Apologies, looks like it breaks the style. Let's revert. Sorry.

**Comment by rootonchair** [comment]
No problem. I have reverted it

**Review by sayakpaul — APPROVED** [review]

**Comment by sayakpaul** [comment]
Thanks for bearing with my requests!

**Pull request merged**